File: hw1/kmeans.py
import sys
import math

EPS = 0.001
DEFAULT_ITER = 400

NUM_CLUST_ERR = "Incorrect number of clusters!"
MAX_ITER_ERR = "Incorrect maximum iteration!"
GENERIC_ERR = "An Error Has Occurred"


def euclid_sq(p: list[float], q: list[float]) -> float:
    """
    Calculate the squared Euclidean distance between two points.

    Args:
        p (list of float): The first point.
        q (list of float): The second point.

    Returns:
        float: The squared Euclidean distance between p and q.
    """
    return sum((pi - qi) ** 2 for pi, qi in zip(p, q))


def add_pts(p: list[float], q: list[float]) -> list[float]:
    """
    Add two points element-wise.

    Args:
        p (list of float): The first point.
        q (list of float): The second point.

    Returns:
        list of float: The element-wise sum of p and q.
    """
    return [pi + qi for pi, qi in zip(p, q)]


def div_pt(p: list[float], denom: float) -> list[float]:
    """
    Divide each element of a point by a denominator.

    Args:
        p (list of float): The point.
        denom (float): The denominator.

    Returns:
        list of float: The point with each element divided by denom.
    """
    return [pi / denom for pi in p]


def parse_args() -> tuple[int, int]:
    """
    Parse command-line arguments for the number of clusters and maximum iterations.

    Returns:
        tuple: A tuple containing the number of clusters (k) and the maximum iterations (max_iter).

    Raises:
        ValueError: If the arguments are invalid.
    """
    if len(sys.argv) not in (2, 3):
        raise ValueError(GENERIC_ERR)

    try:
        k = float(sys.argv[1])
        if not k.is_integer():
            raise ValueError
        k = int(k)
    except ValueError:
        raise ValueError(NUM_CLUST_ERR)

    if len(sys.argv) == 3:
        try:
            iter_ = float(sys.argv[2])
            if not iter_.is_integer():
                raise ValueError
            iter_ = int(iter_)
        except ValueError:
            raise ValueError(MAX_ITER_ERR)
    else:
        iter_ = DEFAULT_ITER

    if not (1 < iter_ < 1000):
        raise ValueError(MAX_ITER_ERR)
    if k < 2:
        raise ValueError(NUM_CLUST_ERR)

    return k, iter_


def read_dataset() -> list[list[float]]:
    """
    Read and parse the dataset from standard input.

    Returns:
        list of list of float: The dataset as a list of points.

    Raises:
        ValueError: If the dataset is invalid.
    """
    data = []
    dim = None
    for raw in sys.stdin.read().splitlines():
        if raw.strip() == "":
            continue
        try:
            point = [float(x) for x in raw.strip().split(",")]
        except ValueError:
            raise ValueError(GENERIC_ERR)

        if dim is None:
            dim = len(point)
        elif len(point) != dim:
            raise ValueError(GENERIC_ERR)

        data.append(point)

    if not data:
        raise ValueError(GENERIC_ERR)
    return data


def kmeans(data: list[list[float]], k: int, max_iter: int) -> list[list[float]]:
    """
    Perform the k-means clustering algorithm.

    Args:
        data (list of list of float): The dataset.
        k (int): The number of clusters.
        max_iter (int): The maximum number of iterations.

    Returns:
        list of list of float: The final centroids of the clusters.

    Raises:
        ValueError: If the number of clusters is invalid.
    """
    n = len(data)
    d = len(data[0])

    if k >= n:
        raise ValueError(NUM_CLUST_ERR)

    #1. Initialize centroids by selecting k first points from the dataset
    centroids = [data[i][:] for i in range(k)]

    for _ in range(max_iter):
        #2. Assign every point to the closest centroid
        clusters = [[] for _ in range(k)]
        for p in data:
            idx = min(range(k), key=lambda i: euclid_sq(p, centroids[i]))
            clusters[idx].append(p)

        #3. Update centroids
        new_centroids = []
        for idx, cluster in enumerate(clusters):
            if cluster:
                acc = [0.0] * d
                for p in cluster:
                    acc = add_pts(acc, p)
                new_centroids.append(div_pt(acc, len(cluster)))
            else:
                new_centroids.append(centroids[idx])

        deltas = [math.sqrt(euclid_sq(nc, oc))
                  for nc, oc in zip(new_centroids, centroids)]
        centroids = new_centroids

        # Update centroids until convergence or max iterations
        if max(deltas) < EPS:
            break


    return centroids


def fmt(c: list[float]) -> str:
    """
    Format a centroid for output.

    Args:
        c (list of float): The centroid.

    Returns:
        str: The formatted centroid as a comma-separated string with 4 decimal places.
    """
    return ",".join(f"{coord:.4f}" for coord in c)


def main() -> None:
    """
    Main function to execute the k-means clustering algorithm.

    Reads input, parses arguments, performs clustering, and prints the results.
    """
    try:
        k, max_iter = parse_args()
        data = read_dataset()
        centroids = kmeans(data, k, max_iter)
        for c in centroids:
            print(fmt(c))
    except ValueError as e:
        print(e)
        sys.exit(1)
    except Exception:
        print(GENERIC_ERR)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: hw1/kmeans.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DEFAULT_ITER 400
#define EPSILON 0.001

/*
 * This program implements the k-means clustering algorithm in C.
 * It reads a dataset from standard input, parses command-line arguments,
 * and performs clustering to find the centroids of the clusters.
 */

/*
 * Structure representing a vector with its data and dimension.
 */
typedef struct
{
    double *data;  /* Array of data points */
    int dimension; /* Dimension of the vector */
} Vector;

/*
 * Structure representing a cluster with its vectors
 */
typedef struct
{
    Vector *data; /* Array of data vectors */
    int size;
    int capacity;
} VectorList;

/*
 * Structure representing a dataset containing multiple vectors.
 */
typedef struct
{
    Vector *vectors; /* Array of vectors */
    int count;       /* Number of vectors in the dataset */
    int dimension;   /* Dimension of the vectors in the dataset */
} Dataset;

/*
 * Parses command-line arguments to extract the number of clusters (K) and maximum iterations.
 *
 * Args:
 *   argc: Number of command-line arguments.
 *   argv: Array of command-line arguments.
 *   K: Pointer to store the number of clusters.
 *   iter: Pointer to store the maximum number of iterations.
 *
 * Returns:
 *   0 on success, 1 on failure.
 */
int parse_arguments(int argc, char *argv[], int *K, int *iter);

/*
 * Reads the dataset from standard input.
 *
 * Returns:
 *   Pointer to the dataset structure, or NULL on failure.
 */
Dataset *read_data(void);

/*
 * Frees the memory allocated for a dataset.
 *
 * Args:
 *   dataset: Pointer to the dataset to free.
 */
void free_dataset(Dataset *dataset);

/*
 * Calculates the Euclidean distance between two vectors.
 *
 * Args:
 *   v1: Pointer to the first vector.
 *   v2: Pointer to the second vector.
 *
 * Returns:
 *   The Euclidean distance between the two vectors.
 */
double euclidean_distance(Vector *v1, Vector *v2);

/*
 * Performs the k-means clustering algorithm.
 *
 * Args:
 *   dataset: Pointer to the dataset.
 *   K: Number of clusters.
 *   max_iter: Maximum number of iterations.
 */
void kmeans(Dataset *dataset, int K, int max_iter);

/*
 * Parses a line of input to create a vector.
 *
 * Args:
 *   line: Input line containing vector data.
 *   expected_dim: Expected dimension of the vector (-1 if unknown).
 *
 * Returns:
 *   Pointer to the created vector, or NULL on failure.
 */
Vector *parse_vector(char *line, int expected_dim);

/*
 * Frees the memory allocated for a vector.
 *
 * Args:
 *   vector: Pointer to the vector to free.
 */
void free_vector(Vector *vector);

int my_strlen(const char *str);

void append_to_cluster(VectorList *cluster, Vector *vec)
{
    if (cluster->size >= cluster->capacity)
    {
        cluster->capacity *= 2;
        cluster->data = realloc(cluster->data, cluster->capacity * sizeof(Vector));
    }
    cluster->data[cluster->size++] = *vec;
}

Dataset *create_sample_dataset(void) {
    Dataset *dataset;
    int i, j;
    double values[4][3] = {
        {1.0, 1.0, 1.0},
        {2.0, 2.0, 2.0},
        {100.0, 100.0, 100.0},
        {101.0, 101.0, 101.0}
    };

    dataset = (Dataset *)malloc(sizeof(Dataset));
    if (dataset == NULL) {
        printf("Allocation failed for dataset\n");
        exit(1);
    }

    dataset->count = 4;
    dataset->dimension = 3;

    dataset->vectors = (Vector *)malloc(dataset->count * sizeof(Vector));
    if (dataset->vectors == NULL) {
        printf("Allocation failed for vectors\n");
        free(dataset);
        exit(1);
    }

    for (i = 0; i < dataset->count; i++) {
        dataset->vectors[i].dimension = dataset->dimension;
        dataset->vectors[i].data = (double *)malloc(dataset->dimension * sizeof(double));

        if (dataset->vectors[i].data == NULL) {
            printf("Allocation failed for vector %d\n", i);
            /* Free all previously allocated memory */
            for (j = 0; j < i; j++) {
                free(dataset->vectors[j].data);
            }
            free(dataset->vectors);
            free(dataset);
            exit(1);
        }

        for (j = 0; j < dataset->dimension; j++) {
            dataset->vectors[i].data[j] = values[i][j];
        }
    }

    return dataset;
}

int main(void)
{
    int K, max_iter;
    Dataset *dataset;
    /*
    *  if (parse_arguments(argc, argv, &K, &max_iter) != 0)
    * {
    *    return 1;
    *}
    */   
    K = 2;
    max_iter = 200;
    dataset = create_sample_dataset();
    if (dataset == NULL)
    {
        printf("An Error Has Occurred\n");
        return 1;
    }
    if (K <= 1 || K >= dataset->count)
    {
        printf("Incorrect number of clusters!\n");
        free_dataset(dataset);
        return 1;
    }
    kmeans(dataset, K, max_iter);
    free_dataset(dataset);
    return 0;
}

int parse_arguments(int argc, char *argv[], int *K, int *iter)
{
    if (argc != 2 && argc != 3)
    {
        printf("An Error Has Occurred\n");
        return 1;
    }
    {
        char *endptr;
        double k_val = strtod(argv[1], &endptr);
        if (endptr == argv[1] || *endptr != '\0' || floor(k_val) != k_val)
        {
            printf("Incorrect number of clusters!\n");
            return 1;
        }
        *K = (int)k_val;
        if (*K < 2)
        {
            printf("Incorrect number of clusters!\n");
            return 1;
        }
    }
    if (argc == 3)
    {
        char *endptr;
        double it_val = strtod(argv[2], &endptr);
        if (endptr == argv[2] || *endptr != '\0' || floor(it_val) != it_val)
        {
            printf("Incorrect maximum iteration!\n");
            return 1;
        }
        *iter = (int)it_val;
        if (*iter <= 1 || *iter >= 1000)
        {
            printf("Incorrect maximum iteration!\n");
            return 1;
        }
    }
    else
    {
        *iter = DEFAULT_ITER;
    }
    return 0;
}

Dataset *read_data(void)
{
    Dataset *dataset;
    char *line = NULL;
    size_t len = 0;
    ssize_t read;
    int capacity = 100;
    int count = 0;
    int dimension = -1;
    Vector *temp_vector;

    dataset = malloc(sizeof(Dataset));
    if (dataset == NULL)
        return NULL;
    dataset->vectors = malloc(capacity * sizeof(Vector));
    if (dataset->vectors == NULL)
    {
        free(dataset);
        return NULL;
    }

    while ((read = getline(&line, &len, stdin)) != -1)
    {
        if (read <= 1)
            continue;
        if (line[read - 1] == '\n')
        {
            line[read - 1] = '\0';
        }
        temp_vector = parse_vector(line, dimension);
        if (temp_vector == NULL)
        {
            free(line);
            free_dataset(dataset);
            return NULL;
        }
        if (dimension == -1)
        {
            dimension = temp_vector->dimension;
        }
        if (temp_vector->dimension != dimension)
        {
            free_vector(temp_vector);
            free(line);
            free_dataset(dataset);
            return NULL;
        }
        if (count >= capacity)
        {
            capacity *= 2;
            dataset->vectors = realloc(dataset->vectors, capacity * sizeof(Vector));
            if (dataset->vectors == NULL)
            {
                free_vector(temp_vector);
                free(line);
                free_dataset(dataset);
                return NULL;
            }
        }
        dataset->vectors[count] = *temp_vector;
        free(temp_vector);
        count++;
    }

    free(line);
    dataset->count = count;
    dataset->dimension = dimension;

    if (count == 0)
    {
        free_dataset(dataset);
        return NULL;
    }
    return dataset;
}

Vector *parse_vector(char *line, int expected_dim)
{
    Vector *vector;
    int dimension = 0;
    int i, j, start, end;
    int line_len;

    line_len = my_strlen(line);
    for (i = 0; i < line_len; i++)
    {
        if (line[i] == ',')
            dimension++;
    }
    dimension++;
    if (dimension == 0)
        return NULL;
    if (expected_dim != -1 && dimension != expected_dim)
        return NULL;
    vector = malloc(sizeof(Vector));
    if (vector == NULL)
        return NULL;
    vector->data = malloc(dimension * sizeof(double));
    if (vector->data == NULL)
    {
        free(vector);
        return NULL;
    }
    vector->dimension = dimension;
    i = 0;
    start = 0;
    for (j = 0; j <= line_len && i < dimension; j++)
    {
        if (j == line_len || line[j] == ',')
        {
            end = j;
            line[end] = '\0';
            vector->data[i] = atof(&line[start]);
            start = j + 1;
            i++;
        }
    }
    return vector;
}

void free_vector(Vector *vector)
{
    if (vector)
    {
        if (vector->data)
            free(vector->data);
        free(vector);
    }
}

void free_dataset(Dataset *dataset)
{
    int i;
    if (dataset)
    {
        if (dataset->vectors)
        {
            for (i = 0; i < dataset->count; i++)
            {
                if (dataset->vectors[i].data)
                {
                    free(dataset->vectors[i].data);
                }
            }
            free(dataset->vectors);
        }
        free(dataset);
    }
}

double euclidean_distance(Vector *v1, Vector *v2)
{
    double sum = 0.0;
    int i;
    for (i = 0; i < v1->dimension; i++)
    {
        double diff = v1->data[i] - v2->data[i];
        sum += diff * diff;
    }
    return sqrt(sum);
}

void kmeans(Dataset *dataset, int K, int max_iter)
{
    Vector *centroids;
    Vector *new_centroids;
    VectorList *clusters;
    int i, j, k, iter;
    double max_change;
    int closest_cluster_index;
    double min_distance, distance;

    centroids = malloc(K * sizeof(Vector));
    new_centroids = malloc(K * sizeof(Vector));
    clusters = malloc(K * sizeof(VectorList));

    if (!centroids || !new_centroids || !clusters)
    {
        printf("An Error Has Occurred\n");
        return;
    }

    /* 1. Initialize centroids by selecting k first points from the dataset */
    for (k = 0; k < K; k++)
    {
        centroids[k].dimension = dataset->dimension;
        centroids[k].data = malloc(dataset->dimension * sizeof(double));
        new_centroids[k].dimension = dataset->dimension;
        new_centroids[k].data = malloc(dataset->dimension * sizeof(double));
        if (!centroids[k].data || !new_centroids[k].data)
        {
            printf("An Error Has Occurred\n");
            return;
        }
        for (j = 0; j < dataset->dimension; j++)
        {
            centroids[k].data[j] = dataset->vectors[k].data[j];
        }
    }

    for (iter = 0; iter < max_iter; iter++)
    {
        /* 2. Assign every point to the closest centroid */
        for (i = 0; i < K; i++)
        {
            clusters[i].size = 0;
            clusters[i].capacity = 4;
            clusters[i].data = malloc(clusters[i].capacity * sizeof(Vector));
        }

        for (i = 0; i < dataset->count; i++)
        {
            min_distance = euclidean_distance(&dataset->vectors[i], &centroids[0]);
            closest_cluster_index = 0;
            for (k = 1; k < K; k++)
            {
                distance = euclidean_distance(&dataset->vectors[i], &centroids[k]);
                if (distance < min_distance)
                {
                    min_distance = distance;
                    closest_cluster_index = k;
                }
            }
            append_to_cluster(&clusters[closest_cluster_index], &dataset->vectors[i]);
        }

        /* 2. Update centeroids */
        for (k = 0; k < K; k++)
        {
            for (j = 0; j < dataset->dimension; j++) {
                new_centroids[k].data[j] = 0.0;
            }
            for (i = 0; i < clusters[k].size; i++)
            {
                for (j = 0; j < dataset->dimension; j++)
                {
                    printf("%f ", clusters[k].data[i].data[j]);
                    new_centroids[k].data[j] += clusters[k].data[i].data[j];
                }
            }
            for (j = 0; j < dataset->dimension; j++)
            {
                new_centroids[k].data[j] /= clusters[k].size;
            }
        }

        for (k = 0; k < K; k++)
        {
            for (j = 0; j < dataset->dimension; j++)
            {
                printf("Value: %f\n", new_centroids[k].data[j]);
                centroids[k].data[j] = new_centroids[k].data[j];
            }
        }

        max_change = 0.0;
        for (k = 0; k < K; k++)
        {
            distance = euclidean_distance(&centroids[k], &new_centroids[k]);
            if (distance > max_change)
            {
                max_change = distance;
            }
        }

        if (max_change < EPSILON)
        {
            break;
        }
    }

    for (k = 0; k < K; k++)
    {
        for (j = 0; j < dataset->dimension; j++)
        {
            if (j > 0)
                printf(",");
            printf("%.4f", centroids[k].data[j]);
        }
        printf("\n");
    }

    for (k = 0; k < K; k++)
    {
        free(centroids[k].data);
        free(new_centroids[k].data);
    }
    free(centroids);
    free(new_centroids);    
}

int my_strlen(const char *str)
{
    int len = 0;
    while (str[len] != '\0')
        len++;
    return len;
}
